m2cgen/assemblers/boosting.py

```diff
@@ -194,6 +195,78 @@ def _count_leaves(tree):
         return num_leaves
 
 
+class XGBoostLinearModelAssembler(ModelAssembler):
```

Comment:
A lot of this code looks a lot like `BaseBoostingAssembler`. Is there a way to generalize this? Let me know if you'd like to discuss this.

Reply:
Yes, at the first glance it looks so. I started from inheriting `BaseBoostingAssembler` and failed to implement new functionality. So, I switched to `ModelAssembler`. The thing is that in `BaseBoostingAssembler` the main attribute and abstraction is `trees` which is passed everywhere. But in linear mode the main abstraction is `weight` and `bias`.

Please let me know if you have a proper way to implement this with `BaseBoostingAssembler` in your mind.

Reply:
I believe we'd have to introduce another entity on top of the `BaseBoostingAssembler`, eg. `BaseTreeBoostingAssembler` and modify the `BaseBoostingAssembler` in a way:
```
class BaseBoostingAssembler(ModelAssembler):
    ...
  def _assemble_estimators(self, *args, **kwargs):
    raise NotImplementedError

  def _assemble_single_output(self, base_score=0, *args, **kwargs):
    estimators_ast = self._assemble_estimators(*args, **kwargs)

    tmp_ast = utils.apply_op_to_expressions(
        ast.BinNumOpType.ADD,
        ast.NumVal(base_score),
        *estimators_ast)

    result_ast = self._final_transform(tmp_ast)

    return ast.SubroutineExpr(result_ast)
```
Same applies to `_assemble_bin_class_output` and so on.
Now for the `BaseTreeBoostingAssembler` the new method should look like following:
```
def _assemble_estimators(self, trees):
  if self._tree_limit:
    trees = trees[:self._tree_limit]

  trees_ast = [self._assemble_tree(t) for t in trees]
  
  # In a large tree we need to generate multiple subroutines to avoid
  # java limitations https://github.com/BayesWitnesses/m2cgen/issues/103.
  trees_num_leaves = [self._count_leaves(t) for t in trees]
  if sum(trees_num_leaves) > self._leaves_cutoff_threshold:
    return self._split_into_subroutines(trees_ast, trees_num_leaves)
  else:
    return trees_ast
```
And for the `XGBoostLinearModelAssembler`:
```
def _assemble_estimators(self, weights, bias):
  coef = utils.to_1d_array(weights)
  return [_linear_to_ast(coef, bias)]
```
Does this make sense? 
I can try to create a PR into your branch if that helps.

Reply:
Yeah, sounds good!

> I can try to create a PR into your branch if that helps.

Please do it! It will help a lot and save much time, as code says more than words 🙂 .

Reply:
Sorry about the delay. I'll try to come up with something over the upcoming weekend.

Reply:
No worries! I don't think we are in a hurry. Please take as much time as you need!